File: 2024/Day2/day2.cc
// Wayne Mack
// Advent of Code
// Day 2

#include <iostream>
#include <fstream>
#include <sstream>
#include <vector>
#include <string>
using namespace std;

vector<vector<int>> readFileToVectors(const string&);
bool isSafe(vector<int>, int);
int findAnswer(vector<vector<int>>, int);
bool Dampener(vector<int>,int);

int main() {
    string filename = "day2.txt"; 
    vector<vector<int>> numbers = readFileToVectors(filename);
    
    cout << endl << "Part One: " << findAnswer(numbers,1);
    cout << endl << "Part Two: " << findAnswer(numbers,2);
    cout << endl;
    return 0;
}

vector<vector<int>> readFileToVectors(const string& filename) {
    vector<vector<int>> result;
    ifstream file(filename);
    // Check if file is open
    if (!file.is_open()) {
        cerr << "Error opening file." << endl;
        return result;  // Return empty result if the file can't be opened
    }
    string line;
    // Read file line by line
    while (getline(file, line)) {
        stringstream ss(line);
        vector<int> lineNumbers;
        int number;
        
        // Read numbers separated by spaces in each line
        while (ss >> number) {
            lineNumbers.push_back(number);
        }
        
        // Add the vector of numbers to the result
        result.push_back(lineNumbers);
    }
    file.close();
    return result;
}

int findAnswer (vector<vector<int>> numericData, int partNumber) {
    int answer = 0;
    for (const auto&  line : numericData) {
        if (isSafe(line,partNumber)) {
            answer++; 
        }
    }         
    return answer;
}

bool isSafe(vector<int> nums, int partNumber) {
    if (nums.size() < 2) {
        return false;
    }
    bool isIncreasing = true;
    bool isDecreasing = true;

    for (size_t i = 1; i < nums.size(); ++i) {
        int diff = nums[i] - nums[i - 1];

        if (diff == 0 || diff > 3 || diff < -3) {
            if (partNumber == 1) {
                return false;            
            } else {
                for (int x = 0; x < nums.size(); x++) {
                    if (Dampener(nums,x)) {
                        return true;
                    }
                }
                return false;  
            }
        }
        if (diff < 0) {
            isIncreasing = false;  
        }
        if (diff > 0) {
            isDecreasing = false; 
        }
        if ((!isDecreasing && !isIncreasing) && (partNumber == 2)) {
           for (int x = 0; x < nums.size(); x++) {
                if (Dampener(nums,x)) {
                        return true;
                }
            }
            return false;  
        }
    }
    // Return true if the sequence is either strictly increasing or strictly decreasing
    return  isIncreasing || isDecreasing;
}

bool Dampener(vector<int> original, int atNumber){
    original.erase(original.begin() + atNumber);
    return isSafe(original,1);
}

File: 2025/day2.py
''' Wayne Mack
    Advent Of Code
    Day 2      '''

import numbers

fileName = "Day2.txt"

def parse_number_file(filename):
    result = []
    with open(filename, 'r') as file:
        content = file.read().strip()
        items = content.split(',')
        for item in items:
            if '-' in item:
                start, end = map(int, item.split('-'))
                result.extend(range(start, end + 1))
            else:
                result.append(int(item))
    return result

numbers = parse_number_file(fileName)

def is_symmetric_split(number, onlyTwo):
    num_str = str(number)
    length = len(num_str)

    if onlyTwo:
        if length % 2 != 0:
            return False
        mid = length // 2
        return num_str[:mid] == num_str[mid:]

    else:
        for i in range(1, length // 2 + 1):
            if length % i == 0:
                part = num_str[:i]
                if part * (length // i) == num_str:
                    return True
        return False



part1Total = 0
part2Total = 0

for n in numbers:
    if is_symmetric_split(n, True):
        part1Total += n
    if is_symmetric_split(n, False):
        part2Total += n

print (part1Total)
print (part2Total)